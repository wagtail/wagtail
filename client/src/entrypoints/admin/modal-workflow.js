/* A framework for modal popups that are loaded via AJAX, allowing navigation to other
subpages to happen within the lightbox, and returning a response to the calling page,
possibly after several navigation steps
*/

import $ from 'jquery';

import { noop } from '../../utils/noop';
import { gettext } from '../../utils/gettext';

/* eslint-disable */
function ModalWorkflow(opts) {
  /* options passed in 'opts':
    'url' (required): URL to the view that will be loaded into the dialog.
      If not provided and dialogId is given, the dialog component's data-url attribute is used instead.
    'dialogId' (optional): the id of the dialog component to use instead of the Bootstrap modal
    'responses' (optional): dict of callbacks to be called when the modal content
      calls modal.respond(callbackName, params)
    'onload' (optional): dict of callbacks to be called when loading a step of the workflow.
      The 'step' field in the response identifies the callback to call, passing it the
      modal object and response data as arguments
  */

  const self = {};
  const responseCallbacks = opts.responses || {};
  const errorCallback = opts.onError || noop;
  const useDialog = !!opts.dialogId;

  if (useDialog) {
    self.dialog = document.getElementById(opts.dialogId);
    self.url = opts.url || self.dialog.dataset.url;
    self.body = self.dialog.querySelector('[data-dialog-body]');

    // Clear the dialog body as it may have been populated previously
    self.body.innerHTML = '';
  } else {
    /* remove any previous modals before continuing (closing doesn't remove them from the dom) */
    $('body > .modal').remove();

    // disable the trigger element so it cannot be clicked twice while modal is loading
    self.triggerElement = document.activeElement;
    self.triggerElement.setAttribute('disabled', true);

    // set default contents of container
    const iconClose =
      '<svg class="icon icon-cross" aria-hidden="true"><use href="#icon-cross"></use></svg>';
    self.container = $(
      '<div class="modal fade" tabindex="-1" role="dialog" aria-hidden="true">\n  <div class="modal-dialog">\n    <div class="modal-content">\n      <button type="button" class="button close button--icon text-replace" data-dismiss="modal">' +
        iconClose +
        gettext('Close') +
        '</button>\n      <div class="modal-body"></div>\n    </div><!-- /.modal-content -->\n  </div><!-- /.modal-dialog -->\n</div>',
    );

    // add container to body and hide it, so content can be added to it before display
    $('body').append(self.container);
    self.container.modal('hide');

    // add listener - once modal is about to be hidden, re-enable the trigger
    self.container.on('hide.bs.modal', () => {
      self.triggerElement.removeAttribute('disabled');
    });

    // add listener - once modal is fully hidden (closed & css transitions end) - re-focus on trigger and remove from DOM
    self.container.on('hidden.bs.modal', function () {
      self.triggerElement.focus();
      self.container.remove();
    });

    self.url = opts.url;
    self.body = self.container.find('.modal-body');
  }

  self.loadUrl = function (url, urlParams) {
    $.get(url, urlParams, self.loadResponseText, 'text').fail(errorCallback);
  };

  self.postForm = function (url, formData) {
    $.post(url, formData, self.loadResponseText, 'text').fail(errorCallback);
  };

  self.ajaxifyForm = function (formSelector) {
    $(formSelector).each(function () {
      const action = this.action;
      if (this.method.toLowerCase() === 'get') {
        $(this).on('submit', function () {
          self.loadUrl(action, $(this).serialize());
          return false;
        });
      } else {
        $(this).on('submit', function () {
          self.postForm(action, $(this).serialize());
          return false;
        });
      }
    });
  };

  self.loadResponseText = function (responseText) {
    const response = JSON.parse(responseText);
    self.loadBody(response);
  };

  self.loadBody = function (response) {
    if (response.html) {
      // if response contains an 'html' item, replace modal body with it
      if (useDialog) {
        self.body.innerHTML = response.html;
      } else {
        self.body.html(response.html);
        self.container.modal('show');
      }
    }

    /* If response contains a 'step' identifier, and that identifier is found in
    the onload dict, call that onload handler */
    if (opts.onload && response.step && response.step in opts.onload) {
      opts.onload[response.step](self, response);
    }
  };

  self.respond = function (responseType) {
    if (responseType in responseCallbacks) {
      const args = Array.prototype.slice.call(arguments, 1);
      responseCallbacks[responseType].apply(self, args);
    }
  };

  self.close = function () {
    if (useDialog) {
      self.dialog.dispatchEvent(new CustomEvent('wagtail:hide'));
    } else {
      self.container.modal('hide');
    }
  };

  self.loadUrl(self.url, opts.urlParams);

  return self;
}
window.ModalWorkflow = ModalWorkflow;
