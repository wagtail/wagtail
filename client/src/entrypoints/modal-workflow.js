/* A framework for modal popups that are loaded via AJAX, allowing navigation to other
subpages to happen within the lightbox, and returning a response to the calling page,
possibly after several navigation steps
*/

import $ from 'jquery';

/* global wagtailConfig */

/* eslint-disable */
function ModalWorkflow(opts) {
  /* options passed in 'opts':
    'url' (required): initial
    'responses' (optional): dict of callbacks to be called when the modal content
      calls modal.respond(callbackName, params)
    'onload' (optional): dict of callbacks to be called when loading a step of the workflow.
      The 'step' field in the response identifies the callback to call, passing it the
      modal object and response data as arguments
  */

  const self = {};
  const responseCallbacks = opts.responses || {};
  // eslint-disable-next-line func-names
  const errorCallback = opts.onError || function () {};

  /* remove any previous modals before continuing (closing doesn't remove them from the dom) */
  $('body > .modal').remove();

  // set default contents of container
  // eslint-disable-next-line max-len
  const iconClose = '<svg class="icon icon-cross" aria-hidden="true" focusable="false"><use href="#icon-cross"></use></svg>';
  // eslint-disable-next-line max-len
  const container = $('<div class="modal fade" tabindex="-1" role="dialog" aria-hidden="true">\n  <div class="modal-dialog">\n    <div class="modal-content">\n      <button type="button" class="button close button--icon text-replace" data-dismiss="modal">' + iconClose + wagtailConfig.STRINGS.CLOSE + '</button>\n      <div class="modal-body"></div>\n    </div><!-- /.modal-content -->\n  </div><!-- /.modal-dialog -->\n</div>');

  // add container to body and hide it, so content can be added to it before display
  $('body').append(container);
  container.modal('hide');

  self.body = container.find('.modal-body');

  // eslint-disable-next-line func-names
  self.loadUrl = function (url, urlParams) {
    $.get(url, urlParams, self.loadResponseText, 'text').fail(errorCallback);
  };

  // eslint-disable-next-line func-names
  self.postForm = function (url, formData) {
    $.post(url, formData, self.loadResponseText, 'text').fail(errorCallback);
  };

  // eslint-disable-next-line func-names
  self.ajaxifyForm = function (formSelector) {
    // eslint-disable-next-line func-names
    $(formSelector).each(function () {
      const action = this.action;
      if (this.method.toLowerCase() === 'get') {
        // eslint-disable-next-line func-names
        $(this).on('submit', function () {
          self.loadUrl(action, $(this).serialize());
          return false;
        });
      } else {
        // eslint-disable-next-line func-names
        $(this).on('submit', function () {
          self.postForm(action, $(this).serialize());
          return false;
        });
      }
    });
  };

  // eslint-disable-next-line func-names
  self.loadResponseText = function (responseText) {
    const response = JSON.parse(responseText);
    self.loadBody(response);
  };

  // eslint-disable-next-line func-names
  self.loadBody = function (response) {
    if (response.html) {
      // if response contains an 'html' item, replace modal body with it
      self.body.html(response.html);
      container.modal('show');
    }

    /* If response contains a 'step' identifier, and that identifier is found in
    the onload dict, call that onload handler */
    if (opts.onload && response.step && (response.step in opts.onload)) {
      opts.onload[response.step](self, response);
    }
  };

  // eslint-disable-next-line func-names
  self.respond = function (responseType) {
    if (responseType in responseCallbacks) {
      // eslint-disable-next-line prefer-rest-params
      const args = Array.prototype.slice.call(arguments, 1);
      responseCallbacks[responseType].apply(self, args);
    }
  };

  // eslint-disable-next-line func-names
  self.close = function () {
    container.modal('hide');
  };

  self.loadUrl(opts.url, opts.urlParams);

  return self;
}
window.ModalWorkflow = ModalWorkflow;
