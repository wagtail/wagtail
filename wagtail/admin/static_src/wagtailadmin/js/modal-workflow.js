/* A framework for modal popups that are loaded via AJAX, allowing navigation to other
subpages to happen within the lightbox, and returning a response to the calling page,
possibly after several navigation steps
*/

function ModalWorkflow(opts) {
    /* options passed in 'opts':
        'url' (required): initial
        'responses' (optional): dict of callbacks to be called when the modal content
            calls modal.respond(callbackName, params)
        'onload' (optional): dict of callbacks to be called when loading a step of the workflow.
            The 'step' field in the response identifies the callback to call, passing it the
            modal object and response data as arguments
    */

    var self = {};
    var responseCallbacks = opts.responses || {};
    var errorCallback = opts.onError || function () {};

    /* remove any previous modals before continuing (closing doesn't remove them from the dom) */
    $('body > .modal').remove();

    // set default contents of container
    var container = $('<div class="modal fade" tabindex="-1" role="dialog" aria-hidden="true">\n    <div class="modal-dialog">\n        <div class="modal-content">\n            <button type="button" class="button close icon text-replace icon-cross" data-dismiss="modal" aria-hidden="true">&times;</button>\n            <div class="modal-body"></div>\n        </div><!-- /.modal-content -->\n    </div><!-- /.modal-dialog -->\n</div>');

    // add container to body and hide it, so content can be added to it before display
    $('body').append(container);
    container.modal('hide');

    self.body = container.find('.modal-body');

    self.loadUrl = function(url, urlParams) {
        $.get(url, urlParams, self.loadResponseText, 'text').fail(errorCallback);
    };

    self.postForm = function(url, formData) {
        $.post(url, formData, self.loadResponseText, 'text').fail(errorCallback);
    };

    self.ajaxifyForm = function(formSelector) {
        $(formSelector).each(function() {
            var action = this.action;
            if (this.method.toLowerCase() == 'get') {
                $(this).on('submit', function() {
                    self.loadUrl(action, $(this).serialize());
                    return false;
                });
            } else {
                $(this).on('submit', function() {
                    self.postForm(action, $(this).serialize());
                    return false;
                });
            }
        });
    };

    self.loadResponseText = function(responseText, textStatus, xhr) {
        var response = JSON.parse(responseText);
        self.loadBody(response);
    };

    self.loadBody = function(response) {
        if (response.html) {
            // if response contains an 'html' item, replace modal body with it
            self.body.html(response.html);
            container.modal('show');
        }

        /* If response contains a 'step' identifier, and that identifier is found in
        the onload dict, call that onload handler */
        if (opts.onload && response.step && (response.step in opts.onload)) {
            opts.onload[response.step](self, response);
        }
    };

    self.respond = function(responseType) {
        if (responseType in responseCallbacks) {
            args = Array.prototype.slice.call(arguments, 1);
            responseCallbacks[responseType].apply(self, args);
        }
    };

    self.close = function() {
        container.modal('hide');
    };

    self.loadUrl(opts.url, opts.urlParams);

    return self;
}
