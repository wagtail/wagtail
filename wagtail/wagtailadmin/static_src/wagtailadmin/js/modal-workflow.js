/* A framework for modal popups that are loaded via AJAX, allowing navigation to other
subpages to happen within the lightbox, and returning a response to the calling page,
possibly after several navigation steps
*/

function ModalWorkflow(opts) {
    /* options passed in 'opts':
        'url' (required): initial
        'responses' (optional): dict of callbacks to be called when the modal content
            calls modal.respond(callbackName, params)
    */

    var self = {};
    var responseCallbacks = opts.responses || {};

    /* remove any previous modals before continuing (closing doesn't remove them from the dom) */
    $('body > .modal').remove();

    // set default contents of container
    var container = $('<div class="modal fade" tabindex="-1" role="dialog" aria-hidden="true">\n    <div class="modal-dialog">\n        <div class="modal-content">\n            <button type="button" class="button close icon text-replace icon-cross" data-dismiss="modal" aria-hidden="true">&times;</button>\n            <div class="modal-body"></div>\n        </div><!-- /.modal-content -->\n    </div><!-- /.modal-dialog -->\n</div>');

    // add container to body and hide it, so content can be added to it before display
    $('body').append(container);
    container.modal('hide');

    self.body = container.find('.modal-body');

    self.loadUrl = function(url, urlParams) {
        $.get(url, urlParams, self.loadResponseText, 'text');
    };

    self.postForm = function(url, formData) {
        $.post(url, formData, self.loadResponseText, 'text');
    };

    self.ajaxifyForm = function(formSelector) {
        $(formSelector).each(function() {
            var action = this.action;
            if (this.method.toLowerCase() == 'get') {
                $(this).submit(function() {
                    self.loadUrl(action, $(this).serialize());
                    return false;
                });
            } else {
                $(this).submit(function() {
                    self.postForm(action, $(this).serialize());
                    return false;
                });
            }
        });
    };

    self.loadResponseText = function(responseText) {
        var response = eval('(' + responseText + ')');

        self.loadBody(response);
    };

    self.loadBody = function(response) {
        if (response.html) {
            // if the response is html
            self.body.html(response.html);
            container.modal('show');
        }

        if (response.onload) {
            // if the response is a function
            response.onload(self);
        }
    };

    self.respond = function(responseType) {
        if (responseType in responseCallbacks) {
            args = Array.prototype.slice.call(arguments, 1);
            responseCallbacks[responseType].apply(self, args);
        }
    };

    self.close = function() {
        container.modal('hide');
    };

    self.loadUrl(opts.url, opts.urlParams);

    return self;
}
