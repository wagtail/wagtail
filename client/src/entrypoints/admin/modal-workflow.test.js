import $ from 'jquery';

window.$ = $;
window.jQuery = $;

import '../../../../wagtail/admin/static_src/wagtailadmin/js/vendor/bootstrap-modal';
import './modal-workflow';

$.get = jest.fn().mockImplementation((url, data, cb) => {
  cb(
    JSON.stringify({ html: `<div id="url">${url}</div>`, data, step: 'start' }),
  );
  return { fail: jest.fn() };
});

describe('modal-workflow', () => {
  beforeEach(() => {
    document.body.innerHTML =
      '<div id="content"><button class="button action-choose" id="trigger">Open</button></div>';
  });

  it('exposes module as global', () => {
    expect(window.ModalWorkflow).toBeDefined();
  });

  it('should generate a modal when called that is removed when closed', () => {
    let modalWorkflow;

    const openModal = () => {
      modalWorkflow = window.ModalWorkflow({ url: 'path/to/endpoint' });
    };

    expect(document.getElementsByClassName('modal')).toHaveLength(0);

    const triggerButton = document.getElementById('trigger');

    triggerButton.addEventListener('click', openModal);
    triggerButton.dispatchEvent(new MouseEvent('click'));

    // check that modal has been created in the DOM
    const modal = document.getElementsByClassName('modal');
    expect(modal).toHaveLength(1);
    expect(modal).toMatchSnapshot();

    // check that modalWorkflow returns self
    expect(modalWorkflow).toBeInstanceOf(Object);
    expect(modalWorkflow).toHaveProperty('close', expect.any(Function));

    // close modal & check it is removed from the DOM
    modalWorkflow.close();
    expect(document.getElementsByClassName('modal')).toHaveLength(0);
  });

  it('should pull focus from the trigger element to the modal and back to the trigger when closed', () => {
    let modalWorkflow;

    const openModal = () => {
      modalWorkflow = window.ModalWorkflow({ url: 'path/to/endpoint' });
    };

    expect(document.getElementsByClassName('modal')).toHaveLength(0);

    const triggerButton = document.getElementById('trigger');

    triggerButton.focus();
    triggerButton.addEventListener('click', openModal);

    expect(document.activeElement).toEqual(triggerButton);
    triggerButton.dispatchEvent(new MouseEvent('click'));

    // check that modal has been created in the DOM & takes focus
    expect(document.getElementsByClassName('modal')).toHaveLength(1);
    expect(document.activeElement.className).toEqual('modal fade in');

    // close modal & check that focus moves back to the trigger
    modalWorkflow.close();
    expect(document.getElementsByClassName('modal')).toHaveLength(0);
    expect(document.activeElement).toEqual(triggerButton);
  });

  it('should disable the trigger element when the modal is opened and re-enable it when closing', () => {
    let modalWorkflow;

    const openModal = () => {
      modalWorkflow = window.ModalWorkflow({ url: 'path/to/endpoint' });
    };

    expect(document.getElementsByClassName('modal')).toHaveLength(0);

    const triggerButton = document.getElementById('trigger');

    triggerButton.focus();
    triggerButton.addEventListener('click', openModal);
    expect(triggerButton.disabled).toBe(false);
    triggerButton.dispatchEvent(new MouseEvent('click'));

    // check that the trigger button is disabled
    expect(triggerButton.disabled).toBe(true);

    // close modal & check that trigger button is re-enabled
    modalWorkflow.close();
    expect(triggerButton.disabled).toBe(false);
  });

  it('should handle onload and URL param options', () => {
    const onload = { start: jest.fn() };
    const urlParams = { page: 23 };

    let modalWorkflow;

    const openModal = () => {
      modalWorkflow = window.ModalWorkflow({
        url: 'path/to/endpoint',
        urlParams,
        onload,
      });
    };

    expect(onload.start).not.toHaveBeenCalled();
    const triggerButton = document.getElementById('trigger');
    triggerButton.addEventListener('click', openModal);

    const event = new MouseEvent('click');
    triggerButton.dispatchEvent(event);

    expect(modalWorkflow).toBeInstanceOf(Object);

    // important: see mock implementation above, returning a response with injected data to validate behaviour
    const response = {
      data: urlParams,
      html: '<div id="url">path/to/endpoint</div>',
      step: 'start',
    };
    expect(onload.start).toHaveBeenCalledWith(modalWorkflow, response);
  });
});
