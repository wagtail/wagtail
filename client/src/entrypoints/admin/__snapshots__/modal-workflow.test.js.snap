// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal-workflow should generate a modal when called that is removed when closed 1`] = `
HTMLCollection [
  <div
    aria-hidden="false"
    class="modal fade in"
    role="dialog"
    tabindex="-1"
  >
    
  
    <div
      class="modal-dialog"
    >
      
    
      <div
        class="modal-content"
      >
        
      
        <button
          class="button close button--icon text-replace"
          data-dismiss="modal"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="icon icon-cross"
            focusable="false"
          >
            <use
              href="#icon-cross"
            />
          </svg>
          Close
        </button>
        
      
        <div
          class="modal-body"
        >
          <div
            id="url"
          >
            path/to/endpoint
          </div>
        </div>
        
    
      </div>
      <!-- /.modal-content -->
      
  
    </div>
    <!-- /.modal-dialog -->
    

  </div>,
]
`;
